src/main/scala/edu/ie3/simona/agent/grid/GridAgentData.scala

```diff
@@ -219,63 +222,84 @@ object GridAgentData {
+          .copy(nodeToReceivedPower = updatedNodeToReceivedPowersMap)
+      )
+    }
+
+    /** Identify and update the vector of already received information.
+      *
+      * @param powerResponse
+      *   Optional power response message
+      * @param nodeToReceived
+      *   Mapping from node uuid to received values
+      * @param senderRef
+      *   Reference of current sender
+      * @param replace
+      *   If existing values may be replaced or not
+      * @return
+      *   The nodal uuid as well as the updated collection of received
+      *   information
+      */
+    private def updateNodalReceivedVector(
+        powerResponse: PowerResponseMessage,
+        nodeToReceived: NodeToReceivedPower,
+        senderRef: ActorRef,
+        replace: Boolean
+    ): Map[UUID, Vector[(ActorRef, Option[PowerResponseMessage])]] = {
+      // extract the nodeUuid that corresponds to the sender's actorRef and check if we expect a message from the sender
+      val nodeUuid = powerResponse match {
+        case powerValuesMessage: ProvidePowerMessage =>
+          uuid(nodeToReceived, senderRef, replace)
+            .getOrElse(
               throw new RuntimeException(
-                s"$actorName Unknown message received. Can't process message $unknownMsg."
+                s"$actorName Received asset power values msg $powerValuesMessage " +
+                  s"from $senderRef which is not in my power values nodes map or which cannot be replaced!"
               )
-          }
-
-          // update the values in the received map
-          val receivedVector = updatedNodeToReceivedPowerValuesMap
+            )
+        case FailedPowerFlow =>
+          uuid(nodeToReceived, senderRef, replace)
             .getOrElse(
-              nodeUuid,
               throw new RuntimeException(
-                s"NodeId $nodeUuid is not part of nodeToReceivedPowerValuesMap!"
+                s"$actorName Received failed power flow message " +
+                  s"from $senderRef which is not in my power values nodes map or which cannot be replaced!"
               )
             )
-            .filterNot { case (k, v) =>
-              k == senderRef &
-                (if (!replace) v.isEmpty else v.isDefined)
-            } :+ (senderRef, powerValueMessageOpt)
+        case unknownMsg =>
+          throw new RuntimeException(
+            s"$actorName Unknown message received. Can't process message $unknownMsg."
+          )
+      }
 
-          updatedNodeToReceivedPowerValuesMap
-            .updated(nodeUuid, receivedVector)
+      // update the values in the received map
+      val receivedVector = nodeToReceived
+        .getOrElse(
+          nodeUuid,
+          throw new RuntimeException(
+            s"NodeId $nodeUuid is not part of nodeToReceivedPowerValuesMap!"
+          )
+        )
+        /* Filter out the entry, that belongs to the given sender and is not defined, yet, in case no replacement shall
```

Comment:
Whenever there is time, just for my better understanding: Could you please give me some insights why this step is necessary?

Reply:
It's actually a good question. We're filtering out the exact value in the vector that we want to replace with the received data here. This is actually only necessary because we use a vector here, a map would take care of this automatically, since each key is mapped to exactly one value. So it'd be a good idea to replace the vector with a map, and this portion of the code can be removed. We can look through this together if you like :)